SplayTree.ml

```diff
@@ -65,34 +110,67 @@ splay_max t = match t with
         | (rl, c, rr) → if rr == leaf
             then ((l, b, rl), c, leaf)
             else match splay_max rr with
-                | leaf          → leaf
-                | (rrl, x, xa) → (((l, b, rl), c, rrl), x, xa)
+                | leaf          → leaf (* TODO: undefined *)
+                | (rrl, x, xa)  → (((l, b, rl), c, rrl), x, xa)
+
+splay_max_noeq ∷ Tree α → Tree α
+splay_max_noeq t = match t with
+    | leaf      → leaf
+    | (l, b, r) → match r with
+        | leaf         → (l, b, leaf)
+        | (rl, c, rr) → match rr with
+            | leaf            →  ((l, b, rl), c, leaf)
+            | (rrl, rrx, rrr) → match splay_max_noeq (rrl, rrx, rrr) with
+                | leaf          → leaf (* TODO: undefined *)
+                | (rrl', x, xa) → (((l, b, rl), c, rrl'), x, xa)
 
 delete ∷ (Ord α, Eq (Tree α)) ⇒ α ⨯ Tree α → Tree α
 delete a t = if t == leaf
-    then leaf
-    else match splay a t with
-        | (l, a', r) → if a == a'
-            then if l == leaf
-                then r
-                else match splay_max l with
-                    | (l', m, r') → (l', m, r)
-            else (l, a', r)
+  then leaf
+  else match splay a t with
+    | leaf       → leaf (* TODO: undefined *)
+    | (l, a', r) → if a == a'
+      then if l == leaf
+        then r
+        else match splay_max l with
+          | leaf        → leaf (* TODO: undefined *)
+          | (l', m, r') → (l', m, r)
+      else (l, a', r)
+
+delete_noeq ∷ Ord α ⇒ α ⨯ Tree α → Tree α
+delete_noeq a t = match t with
+  | leaf         → leaf
+  | (tl, tx, tr) → match splay_noeq a (tl, tx, tr) with
+    | leaf       → leaf (* TODO: undefined *)
+    | (l, a', r) → if a == a'
+      then match l with
+        | leaf         → r
+        | (ll, lx, lr) → match splay_max_noeq (ll, lx, lr) with
+          | leaf          → leaf (* TODO: undefined *)
+          | (ll', m, lr') → (ll', m, lr)
```

Comment:
I think it should be `(ll', m, r)` on the right-hand side here.

Reply:
Nice catch, thank you!

Reply:
I believe that this was resolved by 1b8eecd.

Reply:
you're right